### PR TITLE
Lazily print reprs in log, error, and debugging output

### PR DESCRIPTION
**Pull request by gkaemmer**
Only python for now. Once I get a review, I can port to ruby and potentially node. Other languages (I think) are unlikely to have slow reprs, so it's less important to do this.

PR checklist:

- [x] Added changelog entry.

### CODE REVIEW

File: languages/python/oso/polar/ffi.py [hunk 1]
```diff
@@ -14,70 +17,78 @@ def __del__(self):
...
     def clear_rules(self):
         """Clear all rules from the Polar KB"""
         result = lib.polar_clear_rules(self.ptr)
-        process_messages(self.next_message)
-        check_result(result)
+        self.process_messages()
+        self.check_result(result)
 
-    def new_query_from_str(self, query_str):
+    def new_query_from_str(self, query_str, host):
```
[the remark this answers is not in the document]
Comment (dhatch) [reply languages/python/oso/polar/ffi.py:49]: Nvm I see. Would be good to add a comment about this though.

File: languages/python/oso/polar/ffi.py [hunk 2]
```diff
@@ -14,70 +17,78 @@ def __del__(self):
...
             self.ptr, ffi_serialize(query_term), 0
         )
-        process_messages(self.next_message)
-        query = check_result(new_q_ptr)
-        return Query(query)
+        self.process_messages()
+        query = self.check_result(new_q_ptr)
+        return Query(query, host)
 
     def next_inline_query(self):
```
Comment (dhatch) [inline languages/python/oso/polar/ffi.py:63]: This needs to use the same host class as the host for the inline query.

File: languages/python/oso/polar/polar.py [hunk 3]
```diff
@@ -64,6 +64,7 @@ class Polar:
     def __init__(self, classes=CLASSES):
         self.ffi_polar = FfiPolar()
         self.host = Host(self.ffi_polar)
+        self.ffi_polar.host = self.host
```
Comment (dhatch) [inline languages/python/oso/polar/polar.py:67]: This could be something like `self.ffi_polar.set_message_enricher(callable)` if you want to take my suggestion from above.

File: languages/python/oso/polar/ffi.py [hunk 4]
```diff
@@ -88,43 +99,50 @@ def call_result(self, call_id, value):
...
         # TODO(gj): Do we need to process_messages here?
-        process_messages(self.next_message)
-        check_result(result)
+        self.process_messages()
+        self.check_result(result)
+
+    def check_result(self, result):
+        return check_result(result, lambda msg: self.host.enrich_message(msg))
+
+    def process_messages(self):
```
Comment (dhatch) [inline languages/python/oso/polar/ffi.py:143]: Let's add an assert here that `self.host is not None` to catch forgetting to set the `self.host` property after the constructor. This will help us catch that even if we don't write a test that emits messages.

File: languages/python/oso/tests/test_polar_roles.py [hunk 5]
```diff
@@ -26,6 +26,10 @@ class Org(Base):  # type: ignore
     name = Column(String(), primary_key=True)
 
     def __repr__(self):
+        # TODO: remove this slowdown?
+        import time
```
Comment (dhatch) [inline languages/python/oso/tests/test_polar_roles.py:30]: Remove before merging?

File: languages/python/oso/polar/host.py [hunk 6]
```diff
@@ -190,10 +210,21 @@ def to_polar(self, v):
                     }
                 }
         else:
+            # BEGIN HACK:
```
Comment (dhatch) [inline languages/python/oso/polar/host.py:213]: Thanks for documenting this.

File: languages/python/oso/polar/ffi.py [hunk 7]
```diff
@@ -190,7 +208,10 @@ def process_messages(next_message_method):
         kind = message["kind"]
         msg = message["msg"]
 
+        if postprocess:
```
Comment (dhatch) [inline languages/python/oso/polar/ffi.py:211]: You never used `postprocess`, right?

**Review by dhatch — CHANGES_REQUESTED** [review]
A few small changes.

I left a longer inline comment about this, overall I'm a bit hesitant about introducing this circular dependency. I have some suggestions if you want to take them that could limit the damage a bit but I can't think of another approach that's much better.

File: languages/python/oso/polar/ffi.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gkaemmer) [reply languages/python/oso/polar/ffi.py]: I think this is clearer now that we always use `set_message_enricher`

File: languages/python/oso/polar/ffi.py [hunk 8]
```diff
@@ -1,11 +1,14 @@
 import json
 
 from _polar_lib import ffi, lib
+from polar.host import Host
 
 from .errors import get_python_error
 
 
 class Polar:
+    host: Host
```
[the remark this answers is not in the document]
Comment (gkaemmer) [reply languages/python/oso/polar/ffi.py]: Yeah I think that makes a lot of sense. I removed `host` as an attribute from the `ffi` classes, instead giving them both a `set_message_enricher` method that has to be called before `process_messages` can succeed.

File: languages/python/oso/polar/ffi.py [hunk 9]
```diff
@@ -14,70 +17,78 @@ def __del__(self):
...
-        process_messages(self.next_message)
-        check_result(result)
+        self.process_messages()
+        self.check_result(result)
 
     def next_message(self):
         return lib.polar_next_polar_message(self.ptr)
 
+    def check_result(self, result):
+        return check_result(result, lambda msg: self.host.enrich_message(msg))
```
[the remark this answers is not in the document]
Comment (gkaemmer) [reply languages/python/oso/polar/ffi.py]: TIL -- in node the method would end up unbound to the original `self.host`, but in python that works!

File: languages/python/oso/polar/polar.py [hunk 10]
```diff
@@ -109,6 +109,7 @@ def join(separator, left, right):
                 query = self.ffi_polar.next_inline_query()
                 if query is None:  # Load is done
                     break
+                query.set_message_enricher(self.host.enrich_message)
```
Comment (dhatch) [inline languages/python/oso/polar/polar.py:112]: You'll need to use the copied host made on L114 here. If the inline query creates external instances they are attached to that one instead of `self.host`.

**Comment by gkaemmer** [comment]
@dhatch this is ready for another look. Do you think I should make the ruby/node/etc changes in this PR or in follow-ons? Happy to do either.

File: languages/python/oso/polar/polar.py [hunk 11]
```diff
@@ -151,6 +152,7 @@ def load_str(self, string, filename=None):
             if query is None:  # Load is done
                 break
             else:
+                query.set_message_enricher(self.host.enrich_message)
```
Comment (dhatch) [inline languages/python/oso/polar/polar.py:155]: Same here, use the copied host.

File: languages/python/oso/polar/polar.py [hunk 12]
```diff
@@ -214,9 +218,10 @@ def repl(self, files=[]):
 
             host = self.host.copy()
             host.set_accept_expression(True)
+            ffi_query.set_message_enricher(host.enrich_message)
             result = False
             try:
-                query = Query(ffi_query, host=host).run()
+                query = Query(ffi_query).run()
```
Comment (dhatch) [inline languages/python/oso/polar/polar.py:224]: Why did you remove this `host=`?

File: languages/python/oso/polar/polar.py (hunk 11 above)
Comment (dhatch) [reply languages/python/oso/polar/polar.py:155]: This is really error prone & these would be very hard errors to detect through testing. I think `query.Query` (not `ffi.Query`) can call `ffi.Query.set_message_enricher` in its initializer to ensure that the host classes on the two objects match.

**Review by dhatch — CHANGES_REQUESTED** [review]
Few more things about host

File: languages/python/oso/polar/polar.py (hunk 12 above)
Comment (gkaemmer) [reply languages/python/oso/polar/polar.py:224]: Oh wow good catch -- scary that this doesn't cause any CI failures. I think I got my wires crossed when removing the `host` argument from the `ffi.Query` constructor. Will revert this

File: languages/python/oso/polar/polar.py (hunk 11 above)
Comment (gkaemmer) [reply languages/python/oso/polar/polar.py]: Ah yeah that makes a lot more sense. I just fixed that and its way clearer I think

File: languages/python/oso/polar/polar.py (hunk 12 above)
Comment (dhatch) [reply languages/python/oso/polar/polar.py]: Yeah unfortunately the `host` handling in here is very error prone. (Before your PR it was already a huge rough spot). If you have any thoughts about cleaning it up now that you've looked at it a bit it might be worth us spending time on.

**Review by dhatch — APPROVED** [review]
This looks good to me, can you add a changelog entry before you merge?

Thanks for going through those cleanups! It's fine to do other languages as follow ups in a separate PR (& probably better that way - it's easier to review)
